[Merged by Bors] - Include validator indices in attestation logs

xrchz opened the pull request:
## Issue Addressed

Fixes #2967

## Proposed Changes

Collect validator indices alongside attestations when creating signed
attestations (and aggregates) for inclusion in the logs.

## Additional Info

This is my first time looking at Lighthouse source code and using Rust, so newbie feedback appreciated!

xrchz committed:
Include validator indices in attestation logs

## Issue Addressed

Fixes #2967

## Proposed Changes

Collect validator indices alongside attestations when creating signed
attestations (and aggregates) for inclusion in the logs.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/sigp/lighthouse?pullRequest=3393) <br/>All committers have signed the CLA.

michaelsproul commented on validator_client/src/attestation_service.rs:
```diff
@@ -428,6 +429,7 @@ impl<T: SlotClock + 'static, E: EthSpec> AttestationService<T, E> {
                 log,
                 "Successfully published attestations";
                 "count" => attestations.len(),
+                "validator_indices" => format!("{:?}", validator_indices),
```
Inside the `slog` logging macros use of the `Debug` trait can be abbreviated using `?`

```suggestion
                "validator_indices" => ?validator_indices,
```

michaelsproul commented on validator_client/src/attestation_service.rs:
```diff
@@ -523,11 +525,12 @@ impl<T: SlotClock + 'static, E: EthSpec> AttestationService<T, E> {
+                .await
+                .into_iter()
+                .flatten()
+                .unzip();
```
rather than unzipping here and then re-zipping below we could just collect into a `Vec` of tuples. The `for` loops can still match on `(signed_aggregate_and_proof, validator_index)`

xrchz replied:
if we don't unzip, how would we get the slice on L536?
`
            let signed_aggregate_and_proofs_slice = signed_aggregate_and_proofs.as_slice();
`

michaelsproul replied:
good point, I didn't see that!

michaelsproul reviewed (commented):
Thanks! This is nice quality of life improvement for debugging sub-optimal attestations

Your Rust is impressively idiomatic for a first PR, must be that solid Standard ML background :wink:

xrchz committed:
Use `slog`'s `?` abbreviation for the `Debug` trait

Co-authored-by: Michael Sproul <micsproul@gmail.com>

michaelsproul commented on validator_client/src/attestation_service.rs:
```diff
@@ -545,11 +548,15 @@ impl<T: SlotClock + 'static, E: EthSpec> AttestationService<T, E> {
                             log,
-                            "Successfully published attestations";
+                            "Successfully published attestation";
+                            "validator_index" => validator_index,
```
On closer inspection I think maybe we don't need to add anything to this log, the validator index is already present as `aggregator`

xrchz replied:
Thanks. Glad you noticed this! Done.

michaelsproul commented on validator_client/src/attestation_service.rs:
```diff
@@ -560,12 +567,16 @@ impl<T: SlotClock + 'static, E: EthSpec> AttestationService<T, E> {
                             log,
                             "Failed to publish attestation";
                             "error" => %e,
+                            "validator_index" => validator_index,
```
Here we can also grab the validator index from `signed_aggregate_and_proof.message.aggregator_index`

xrchz replied:
Done.

michaelsproul reviewed (changes requested):
Sorry, I was about to approve this when I realised we'd be duplicating info by adding the validator index to the aggregate attestation log

xrchz committed:
Undo most changes for aggregated attestations

The validator index for aggregated attestations is available as part of
the signed aggregate message, so no extra plumbing is required.

michaelsproul reviewed (approved):
Awesome thanks!

michaelsproul commented:
bors r+

michaelsproul commented:
CI being flaky

bors retry
